File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/entity/component/ScriptComponentActor.java
package net.bestia.zoneserver.actor.entity.component;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.PoisonPill;
import akka.actor.Terminated;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import net.bestia.entity.EntityService;
import net.bestia.entity.component.ScriptComponent;
import net.bestia.messages.EntityComponentUpdateMessage;
import net.bestia.zoneserver.actor.SpringExtension;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.Objects;
import java.util.Set;

/**
 * Manages the {@link ScriptComponent} for an entity.
 *
 * @author Thomas Felix
 */
@Component
@Scope("prototype")
public class ScriptComponentActor extends AbstractActor {

	public final static String NAME = "scriptComponent";

	private final LoggingAdapter LOG = Logging.getLogger(getContext().system(), this);

	private final long entityId;
	private final long componentId;
	private final EntityService entityService;

	private final BiMap<String, ActorRef> scriptActors = HashBiMap.create();

	@Autowired
	public ScriptComponentActor(long entityId, long componentId, EntityService entityService) {
		this.entityId = entityId;
		this.componentId = componentId;
		this.entityService = Objects.requireNonNull(entityService);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder()
				.match(EntityComponentUpdateMessage.class, this::handleComponentUpdate)
				.match(Terminated.class, this::handleTerminated)
				.build();
	}

	/**
	 * The script component has changed. We try to look into the changes and create or delete all
	 * the actors managing the components.
	 *
	 * @param msg
	 */
	private void handleComponentUpdate(EntityComponentUpdateMessage msg) {
		entityService.getComponent(componentId, ScriptComponent.class).ifPresent(c -> {
			final Set<String> currentActiveUids = scriptActors.keySet();
			final Set<String> componentActiveUids = c.getAllScriptUids();

			componentActiveUids.stream()
					.filter(cuid -> componentActiveUids.contains(cuid))
					.map(c::getCallback)
					.forEach(this::updateActor);

			currentActiveUids.stream()
					.filter(cuid -> !componentActiveUids.contains(cuid))
					.forEach(this::terminateActor);

			componentActiveUids.stream()
					.filter(cuid -> !currentActiveUids.contains(cuid))
					.map(c::getCallback)
					.forEach(this::createActor);
		});
	}

	private void terminateActor(String uid) {
		LOG.debug("Periodic script actor terminating: {}", uid);
		final ActorRef ref = scriptActors.get(uid);
		if (ref != null) {
			ref.tell(PoisonPill.getInstance(), getSelf());
		}
	}

	private void createActor(ScriptComponent.ScriptCallback callback) {
		final ActorRef actor = SpringExtension.unnamedActorOf(getContext(),
				PeriodicScriptActor.class, entityId,
				callback.getIntervalMs(), callback.getScript());

		getContext().watch(actor);
	}

	private void updateActor(ScriptComponent.ScriptCallback callback) {
		final ActorRef ref = scriptActors.get(callback.getUuid());
		if(ref != null) {
			ref.tell(callback, getSelf());
		}
	}

	/**
	 * Handle the termination of the periodic movement and remove the actor ref
	 * so we can start a new one.
	 */
	private void handleTerminated(Terminated term) {
		LOG.debug("Periodic script actor terminated: {}", term.getActor().toString());

		final ActorRef termActor = term.actor();
		scriptActors.inverse().remove(termActor);
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/entity/component/ScriptComponentActor.kt
package net.bestia.zoneserver.actor.entity.component

import akka.actor.ActorRef
import akka.actor.PoisonPill
import akka.actor.Terminated
import akka.japi.pf.ReceiveBuilder
import com.google.common.collect.HashBiMap
import mu.KotlinLogging
import net.bestia.messages.entity.EntityMessage
import net.bestia.zoneserver.actor.ActorComponent
import net.bestia.zoneserver.actor.SpringExtension
import net.bestia.zoneserver.entity.component.IntervalScriptCallback
import net.bestia.zoneserver.entity.component.ScriptCallback
import net.bestia.zoneserver.entity.component.ScriptComponent
import java.lang.IllegalStateException
import java.time.Duration

private val LOG = KotlinLogging.logger { }

data class SetTimeoutCommand(
    override val entityId: Long,
    val timeout: Duration,
    val callbackFn: String
) : EntityMessage, ComponentMessage<ScriptComponent> {
  override val componentType: Class<out ScriptComponent>
    get() = ScriptComponent::class.java
}

data class SetIntervalCommand(
    override val entityId: Long,
    val uuid: String,
    val timeout: Duration,
    val callbackFn: String
) : EntityMessage, ComponentMessage<ScriptComponent> {
  override val componentType: Class<out ScriptComponent>
    get() = ScriptComponent::class.java
}

/**
 * Manages the [ScriptComponent] for an entity.
 *
 * @author Thomas Felix
 */
@ActorComponent(ScriptComponent::class)
class ScriptComponentActor(
    scriptComponent: ScriptComponent
) : ComponentActor<ScriptComponent>(scriptComponent) {

  private val periodicScriptActor = HashBiMap.create<String, ActorRef>()

  override fun createReceive(builder: ReceiveBuilder) {
    builder
        .match(Terminated::class.java, this::handlePeriodicActorTerminated)
        .match(SetIntervalCommand::class.java, this::addInterval)
        .match(SetTimeoutCommand::class.java, this::addTimeout)
  }

  private fun addTimeout(msg: SetTimeoutCommand) {
    throw IllegalStateException("No implemented")
  }

  private fun addInterval(msg: SetIntervalCommand) {
    val intervalCallback = IntervalScriptCallback(
        uuid = msg.uuid,
        interval = msg.timeout,
        scriptKeyCallback = msg.callbackFn,
        scriptEntityId = component.entityId
    )
    addPeriodicScriptActor(intervalCallback)
  }

  private fun handlePeriodicActorTerminated(msg: Terminated) {
    LOG.debug { "PeriodicScriptActor ${msg.actor.path()} terminated" }
    periodicScriptActor.inverse().remove(msg.actor)
  }

  override fun onComponentChanged(oldComponent: ScriptComponent, newComponent: ScriptComponent) {
    val periodicScriptActorsToRemove = periodicScriptActor.keys - component.scripts.keys
    periodicScriptActorsToRemove.forEach { killPeriodicActor(it) }
  }

  private fun killPeriodicActor(key: String) {
    periodicScriptActor[key]?.tell(PoisonPill.getInstance(), self)
  }

  private fun addPeriodicScriptActor(scriptCallback: ScriptCallback) {
    LOG.debug { "PeriodicScriptActor $scriptCallback added" }
    val periodicActor = SpringExtension.actorOf(context, IntervalScriptActor::class.java, scriptCallback)
    periodicScriptActor[scriptCallback.uuid] = periodicActor
  }

  companion object {
    const val NAME = "scriptComponent"
  }
}
